framework/src/base/Assembly.C

```diff
@@ -3348,8 +3347,10 @@ Assembly::cacheResidual()
 void
 Assembly::cacheResidual(dof_id_type dof, Real value, TagID tag_id)
 {
-  _cached_residual_values[tag_id].push_back(value);
-  _cached_residual_rows[tag_id].push_back(dof);
+  const VectorTag & tag = _subproblem.getVectorTag(tag_id);
```

Comment:
You may want to add in the header explaining that the tag id passed in here is not the _type_id of a tag.

Reply:
Or make this unchanged and change the other few places calling this function. I think this makes things cleaner in sense that all the TagID in assembly are _type_id.

Reply:
API should always use  tag_id. We only index local vectors using _type_id.